### PR TITLE
[GlobalISel] Bail out early for big-endian

### LINKED ISSUE #103032: [AArch64] Crash in call lowering for function taking vector of non-p2 integer type on big endian
```llvm
target datalayout = "E-m:e-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128"
target triple = "aarch64_be-unknown-linux-musl"

define fastcc void @foo(<4 x i6> %0) {
  ret void
}
```

```console
❯ llc -version
LLVM (http://llvm.org/):
  LLVM version 18.1.8
❯ llc -O0 reduced.ll
llc: /home/alexrp/Source/tests/llvm-project-18/llvm/lib/CodeGen/GlobalISel/CallLowering.cpp:497: void buildCopyFromRegs(llvm::MachineIRBuilder&, llvm::ArrayRef<llvm::Register>, llvm::ArrayRef<llvm::Register>, llvm::LLT, llvm::LLT, llvm::ISD::ArgFlagsTy): Assertion `(SrcEltTy.getSizeInBits() % OriginalEltTy.getSizeInBits()) == 0' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: llc -O0 reduced.ll
1.      Running pass 'Function Pass Manager' on module 'reduced.ll'.
2.      Running pass 'IRTranslator' on function '@foo'
 #0 0x00007e9ecc75d8e2 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/opt/llvm/bin/../lib/libLLVM.so.18.1+0xd5d8e2)
 #1 0x00007e9ecc75a8ff llvm::sys::RunSignalHandlers() (/opt/llvm/bin/../lib/libLLVM.so.18.1+0xd5a8ff)
 #2 0x00007e9ecc75aa45 SignalHandler(int) Signals.cpp:0:0
 #3 0x00007e9ecb242990 (/lib/x86_64-linux-gnu/libc.so.6+0x42990)
 #4 0x00007e9ecb299a1b __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #5 0x00007e9ecb299a1b __pthread_kill_internal ./nptl/pthread_kill.c:78:10
 #6 0x00007e9ecb299a1b pthread_kill ./nptl/pthread_kill.c:89:10
 #7 0x00007e9ecb2428e6 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #8 0x00007e9ecb2268b7 abort ./stdlib/abort.c:81:7
 #9 0x00007e9ecb2267db _nl_load_domain ./intl/loadmsgcat.c:1177:9
#10 0x00007e9ecb239206 (/lib/x86_64-linux-gnu/libc.so.6+0x39206)
#11 0x00007e9ecd5e12a9 buildCopyFromRegs(llvm::MachineIRBuilder&, llvm::ArrayRef<llvm::Register>, llvm::ArrayRef<llvm::Register>, llvm::LLT, llvm::LLT, llvm::ISD::ArgFlagsTy) CallLowering.cpp:0:0
#12 0x00007e9ecd5e4ed7 llvm::CallLowering::handleAssignments(llvm::CallLowering::ValueHandler&, llvm::SmallVectorImpl<llvm::CallLowering::ArgInfo>&, llvm::CCState&, llvm::SmallVectorImpl<llvm::CCValAssign>&, llvm::MachineIRBuilder&, llvm::ArrayRef<llvm::Register>) const (/opt/llvm/bin/../lib/libLLVM.so.18.1+0x1be4ed7)
#13 0x00007e9ecf3ae0ba llvm::AArch64CallLowering::lowerFormalArguments(llvm::MachineIRBuilder&, llvm::Function const&, llvm::ArrayRef<llvm::ArrayRef<llvm::Register>>, llvm::FunctionLoweringInfo&) const (/opt/llvm/bin/../lib/libLLVM.so.18.1+0x39ae0ba)
#14 0x00007e9ecd64d8d5 llvm::IRTranslator::runOnMachineFunction(llvm::MachineFunction&) (/opt/llvm/bin/../lib/libLLVM.so.18.1+0x1c4d8d5)
#15 0x00007e9eccd271f6 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (.part.0) MachineFunctionPass.cpp:0:0
#16 0x00007e9ecc94f4d1 llvm::FPPassManager::runOnFunction(llvm::Function&) (/opt/llvm/bin/../lib/libLLVM.so.18.1+0xf4f4d1)
#17 0x00007e9ecc94f8e1 llvm::FPPassManager::runOnModule(llvm::Module&) (/opt/llvm/bin/../lib/libLLVM.so.18.1+0xf4f8e1)
#18 0x00007e9ecc950207 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/opt/llvm/bin/../lib/libLLVM.so.18.1+0xf50207)
#19 0x0000654acacfc896 compileModule(char**, llvm::LLVMContext&) llc.cpp:0:0
#20 0x0000654acacf1117 main (/opt/llvm/bin/llc+0x12117)
#21 0x00007e9ecb228150 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:74:3
#22 0x00007e9ecb228209 call_init ./csu/../csu/libc-start.c:128:20
#23 0x00007e9ecb228209 __libc_start_main ./csu/../csu/libc-start.c:347:5
#24 0x0000654acacf1e15 _start (/opt/llvm/bin/llc+0x12e15)
```

### PR DESCRIPTION
**Pull request by davemgreen**
If we continue through the function we can currently hit crashes. We can bail out early and fall back to SDAG.

Fixes #103032

### CODE REVIEW

File: llvm/test/CodeGen/AArch64/GlobalISel/endian_fallback.ll [hunk 1]
```diff
@@ -0,0 +1,18 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple aarch64-unknown-linux-musl -O0 < %s | FileCheck %s --check-prefix=CHECK-LE
+; RUN: llc -mtriple aarch64_be-unknown-linux-musl -O0 < %s | FileCheck %s --check-prefix=CHECK-BE
+
+; Make sure we fall-back to SDAG for BE targets.
```
Comment (arsenm) [inline llvm/test/CodeGen/AArch64/GlobalISel/endian_fallback.ll:5]: This doesn't really do that unless you check for the fallback remark

**Comment by tschuett** [comment]
Silly question: why was "unable to translate in big endian mode" not visible in the error report?

**Comment by davemgreen** [comment]
It looks like this is failing on a couple of Arm tests because it is expecting the ir-translator to produce testable results even under BE.

> Silly question: why was "unable to translate in big endian mode" not visible in the error report?

By "error report", do you mean the fall-back remarks? -O0 will use -global-isel-abort=0 (I think), so not report the remarks or error out, it just prints a dbgs() message and otherwise silently falls back to SDAG.

**Comment by tschuett** [comment]
> It looks like this is failing on a couple of Arm tests because it is expecting the ir-translator to produce testable results even under BE.
> 
> > Silly question: why was "unable to translate in big endian mode" not visible in the error report?
> 
> By "error report", do you mean the fall-back remarks? -O0 will use -global-isel-abort=0 (I think), so not report the remarks or error out, it just prints a dbgs() message and otherwise silently falls back to SDAG.

The message was not visible in https://github.com/llvm/llvm-project/issues/103032. You put a return after the remark. I would have expected the remark to visible.

**Comment by davemgreen** [comment]
Yeah it will depend on the value of -global-isel-abort, which I believe is =0 when enabled through -O0. https://godbolt.org/z/49Pav884n

**Comment by davemgreen** [comment]
I've added an option for testing the Arm backend under BE.

File: llvm/test/CodeGen/AArch64/GlobalISel/endian_fallback.ll (hunk 1 above)
Comment (davemgreen) [reply llvm/test/CodeGen/AArch64/GlobalISel/endian_fallback.ll:5]: Sorry - I missed your remark earlier, my bad. It was crashing earlier so was fairly obvious that it was "fixed". Let me add a check for the remark too.

**Review by aemerson — APPROVED** [review]

**Pull request merged**